diff --git a/build-aux/prefix-gnulib-mk b/build-aux/prefix-gnulib-mk
index 347ddcd64..4eea8b738 100755
--- a/build-aux/prefix-gnulib-mk
+++ b/build-aux/prefix-gnulib-mk
@@ -40,6 +40,7 @@ use File::Basename; # for dirname
 (my $ME = $0) =~ s|.*/||;
 
 my $prefix;
+my $varprefix;
 my $lib_name;
 
 sub usage ($)
@@ -162,7 +163,7 @@ sub prefix_assignment ($$)
 
   # Variables whose name depend on the location: libbison_a_SOURCES =>
   # lib_libbison_a_SOURCES.
-  $lhs_and_assign_op =~ s/($lib_name)/lib_$1/g;
+  $lhs_and_assign_op =~ s/($lib_name)/${varprefix}$1/g;
 
   $lhs_and_assign_op . $rhs;
 }
@@ -243,6 +244,8 @@ sub process ($)
 
   my $file = $ARGV[0];
   $prefix = (dirname $file) . '/';
+  $varprefix = $prefix;
+  $varprefix =~ s{/}{_}g;
   warn "prefix=$prefix\n";
 
   process $file;
